File: app/src/main/java/com/kite/mnemoai/ui/vocabularybookgroup/VocabularyGroupViewModel.java
package com.kite.mnemoai.ui.vocabularybookgroup;

import static androidx.lifecycle.ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY;
import static androidx.lifecycle.SavedStateHandleSupport.createSavedStateHandle;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.SavedStateHandle;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.viewmodel.ViewModelInitializer;

import com.kite.mnemoai.MainApplication;
import com.kite.mnemoai.data.local.entity.GroupEntity;
import com.kite.mnemoai.data.model.WordListItem;
import com.kite.mnemoai.data.repository.GroupRepository;
import com.kite.mnemoai.data.repository.WordRepository;

import java.util.List;
import java.util.stream.Collectors;

import javax.inject.Inject;

import dagger.hilt.android.lifecycle.HiltViewModel;

@HiltViewModel
public class VocabularyGroupViewModel extends ViewModel {
    private MediatorLiveData<VocabularyGroupUIState> uiState = new MediatorLiveData<>();
    private WordRepository wordRepository;
    private GroupRepository groupRepository;
    private GroupEntity group;
    private List<WordListItem> words;

    @Inject
    public VocabularyGroupViewModel(WordRepository wordRepository, GroupRepository groupRepository, SavedStateHandle savedStateHandle) {
        this.wordRepository = wordRepository;
        this.groupRepository = groupRepository;
        long groupId = 0;
        if(savedStateHandle.contains("group_id") && savedStateHandle.get("group_id") != null){
             groupId = (long) savedStateHandle.get("group_id");
        }

        uiState.addSource(groupRepository.getGroupLiveDataById(groupId), (group) -> {
            this.group = group;
            updateUIState();
        });
        uiState.addSource(wordRepository.getWordListItemLiveDataById(groupId), (words) -> {
            this.words = words.stream().peek(word -> {
                word.setPhonetic("/" + word.getPhonetic() + "/");
                word.setTranslation(word.getTranslation().replace("\\n", " "));
            }).collect(Collectors.toList());
            updateUIState();
        });
    }

    public void updateUIState(){
        uiState.setValue(new VocabularyGroupUIState(group, words));
    }

    public GroupEntity getGroup() {
        return group;
    }

    public void modifyVocabularyBook(String name, String desc){
        groupRepository.modifyVocabularyBook(new GroupEntity(group.getId(), name, desc, group.isLearning(), group.getCreateTime()));
    }

    public void setVocabularyBookLearningStatus(int status){
        groupRepository.setVocabularyBookLearningStatus(group.getId(), status);
    }

    public void deleteGroup(){
        groupRepository.deleteGroup(group.getId());
    }

    public LiveData<VocabularyGroupUIState> getUiState(){
        return uiState;
    }

}

File: feature/vocabularybookgroup/src/main/java/com/kite/mnemoai/vocabularybookgroup/VocabularyGroupViewModel.kt
package com.kite.mnemoai.vocabularybookgroup

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.kite.mnemoai.model.Result
import com.kite.mnemoai.model.group.Group
import com.kite.mnemoai.model.repository.GroupRepository
import com.kite.mnemoai.model.repository.WordRepository
import com.kite.mnemoai.model.word.WordItem
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class VocabularyGroupViewModel @Inject constructor(
    private val wordRepository: WordRepository,
    private val groupRepository: GroupRepository,
    savedStateHandle: SavedStateHandle,
) : ViewModel() {

    private val _uiState = MutableLiveData<VocabularyGroupUIState>()
    val uiState: LiveData<VocabularyGroupUIState> get() = _uiState

    private var group: Group? = null
    private var words: List<WordItem>? = null

    init {
        var groupId: Long = 0
        if (savedStateHandle.contains("group_id") && savedStateHandle.get<Long?>("group_id") != null) {
            groupId = savedStateHandle.get<Long?>("group_id") ?: 0
        }

        viewModelScope.launch {
            groupRepository.observeGroupById(groupId).collect { result ->
                if (result is Result.Success) {
                    group = result.data
                    updateUIState()
                }
            }
        }
        viewModelScope.launch {
            wordRepository.observeWordListByGroupId(groupId).collect { result ->
                if (result is Result.Success) {
                    words = result.data.map { word ->
                        word.copy(
                            phonetic = "/${word.phonetic}/",
                            translation = word.translation.replace("\\n", " ")
                        )
                    }
                    updateUIState()
                }
            }
        }
    }

    private fun updateUIState() {
        val list = words.orEmpty()
        _uiState.value = VocabularyGroupUIState(
            group = group,
            words = words,
            allCount = list.size.toLong(),
            learningCount = list.count { it.reviewState == 0 }.toLong(),
            reviewingCount = list.count { it.reviewState == 1 }.toLong(),
            masteredCount = list.count { it.reviewState == 2 }.toLong(),
        )
    }

    fun getGroup(): Group? = group

    fun modifyVocabularyBook(name: String, desc: String) {
        val currentGroup = group ?: return
        viewModelScope.launch {
            groupRepository.modifyVocabularyBook(
                currentGroup.copy(name = name, description = desc)
            )
        }
    }

    fun setVocabularyBookLearningStatus(status: Int) {
        val currentGroup = group ?: return
        viewModelScope.launch {
            groupRepository.setVocabularyBookLearningStatus(currentGroup.id, status)
        }
    }

    fun deleteGroup() {
        val currentGroup = group ?: return
        viewModelScope.launch {
            groupRepository.deleteGroup(currentGroup.id)
        }
    }
}